Mathlib/NumberTheory/FLT/Three.lean

```diff
@@ -3,38 +3,130 @@ Copyright (c) 2024 Riccardo Brasca. All rights reserved.
   rcases cube_of_not_dvd hdvd.1.2 with hb | hb <;>
   rcases cube_of_not_dvd hdvd.2 with hc | hc <;>
   rw [ha, hb, hc] <;> decide
+
+end case1
+
+section case2
+
+private lemma three_dvd_c_of_dvd_a_of_dvd_b {a b c : ℤ} (HF : a ^ 3 + b ^ 3 + c ^ 3 = 0)
+    (h3a : 3 ∣ a) (h3b : 3 ∣ b) : 3 ∣ c := by
+  refine Int.prime_three.dvd_of_dvd_pow (n := 3) (dvd_neg.1 ?_)
+  rw [add_eq_zero_iff_eq_neg] at HF
+  exact HF.symm ▸ dvd_add (dvd_pow h3a (by decide)) (dvd_pow h3b (by decide))
+
+private lemma three_dvd_b_of_dvd_a_of_gcd_eq_one_of_case2 {a b c : ℤ} (ha : a ≠ 0) (hc : c ≠ 0)
+    (Hgcd: Finset.gcd {a, b, c} id = 1) (h3a : 3 ∣ a) (HF : a ^ 3 + b ^ 3 + c ^ 3 = 0)
+    (H : ∀ a b c : ℤ, c ≠ 0 → ¬ 3 ∣ a → ¬ 3 ∣ b  → 3 ∣ c → IsCoprime a b → a ^ 3 + b ^ 3 ≠ c ^ 3) :
+    3 ∣ b := by
+  have hbc : IsCoprime (-b) (-c) := by
+    refine IsCoprime.neg_neg <| isCoprime_of_prime_dvd  ?_<| (fun p hp hpb hpc ↦ hp.not_dvd_one ?_)
+    · simp [hc]
+    · rw [← Hgcd]
+      refine Finset.dvd_gcd_iff.2 (fun x hx ↦ ?_)
+      simp only [Finset.mem_insert, Finset.mem_singleton] at hx
+      rcases hx with (hx | hx | hx)
+      · simp only [hx, id_eq]
+        refine hp.dvd_of_dvd_pow (n := 3) ?_
+        rw [add_eq_zero_iff_eq_neg, ← eq_sub_iff_add_eq] at HF
+        exact HF ▸ dvd_sub (dvd_neg.2 (dvd_pow hpc (by decide))) (dvd_pow hpb (by decide))
+      · exact hx ▸ hpb
+      · exact hx ▸ hpc
+  by_contra! h3b
+  by_cases h3c : 3 ∣ c
+  · apply h3b
+    rw [add_assoc, add_comm (b ^ 3), ← add_assoc] at HF
+    exact three_dvd_c_of_dvd_a_of_dvd_b HF h3a h3c
+  · refine H (-b) (-c) a ha (by simp [h3b]) (by simp [h3c]) h3a hbc ?_
+    rw [add_eq_zero_iff_eq_neg, ← (show Odd 3 by decide).neg_pow] at HF
+    rw [← HF]
+    ring
+
+open Finset Int in
+
+/--
+  For all `a`, `b`, `c` in `ℤ` such that `c ≠ 0`, `¬ 3 ∣ a`, `¬ 3 ∣ b`, `a` and `b`
+are coprime and `3 ∣ c`, then `a ^ 3 + b ^ 3 ≠ c ^ 3`.
+-/
+theorem fermatLastTheoremThree_of_three_dvd_only_c
+    (H : ∀ a b c : ℤ, c ≠ 0 → ¬ 3 ∣ a → ¬ 3 ∣ b  → 3 ∣ c → IsCoprime a b → a ^ 3 + b ^ 3 ≠ c ^ 3) :
+    FermatLastTheoremFor 3 := by
+  rw [fermatLastTheoremFor_iff_int]
+  refine fermatLastTheoremWith_of_fermatLastTheoremWith_coprime (fun a b c ha hb hc Hgcd hF ↦?_)
+  by_cases h1 : 3 ∣ a * b * c
+  swap; exact fermatLastTheoremThree_case1 h1 hF
+  rw [(prime_three).dvd_mul, (prime_three).dvd_mul] at h1
+  apply (show ¬(3 ∣ (1 : ℤ)) by decide)
+  rw [← Hgcd]
+  rw [← sub_eq_zero, sub_eq_add_neg, ← (show Odd 3 by decide).neg_pow] at hF
+  refine dvd_gcd (fun x hx ↦ ?_)
+  simp only [mem_insert, mem_singleton] at hx
+  rcases h1 with ((h3a | h3b) | h3c)
+  · have h3b : 3 ∣ b := by
+      refine three_dvd_b_of_dvd_a_of_gcd_eq_one_of_case2 ha (neg_ne_zero.2 hc) ?_ h3a hF H
+      simp only [← Hgcd, gcd_insert, gcd_singleton, id_eq, ← abs_eq_normalize, abs_neg]
+    rcases hx with (hx | hx | hx)
+    · exact hx ▸ h3a
+    · exact hx ▸ h3b
+    · simpa [hx] using three_dvd_c_of_dvd_a_of_dvd_b hF h3a h3b
```

Comment:
This block is still essentially repeated three times. I think it makes sense to extract one of the cases (say, `3 ∣ a`) as a separate lemma (with roughly the six lines above as its proof) and then use this helper lemma three times.